cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,288 @@
+	kubeClient      clientset.Interface
+	client          *godo.Client
+	tags            []string
+	serviceLister   corelisters.ServiceLister
+	firewallManager *FirewallManager
+	fwCache         *firewallCache
+	ccmFirewallName string
+}
+
+// NewFirewallController returns a new firewall controller to reconcile CCM worker firewall state.
+func NewFirewallController(
+	kubeClient clientset.Interface,
+	client *godo.Client,
+	serviceInformer coreinformers.ServiceInformer,
+	tags []string,
+	fwManager *FirewallManager,
+	ctx context.Context,
+) (*FirewallController, error) {
+	if kubeClient != nil && kubeClient.CoreV1().RESTClient().GetRateLimiter() != nil {
+		if err := ratelimiter.RegisterMetricAndTrackRateLimiterUsage("firewall_controller", kubeClient.CoreV1().RESTClient().GetRateLimiter()); err != nil {
+			return nil, err
+		}
+	}
+
+	fc := &FirewallController{
+		kubeClient:      kubeClient,
+		client:          client,
+		firewallManager: fwManager,
+		fwCache: &firewallCache{
+			mu: new(sync.RWMutex),
+		},
+	}
+
+	serviceInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				fc.onServiceChange(ctx)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				fc.onServiceChange(ctx)
+			},
+			DeleteFunc: func(obj interface{}) {
+				fc.onServiceChange(ctx)
+			},
+		},
+		serviceSyncPeriod,
+	)
+	fc.serviceLister = serviceInformer.Lister()
+
+	return fc, nil
+}
+
+// Run starts the firewall controller loop.
+func (fc *FirewallController) Run(stopCh <-chan struct{}) {
+	ctx, cancel := context.WithTimeout(context.Background(), serviceSyncPeriod)
+	defer cancel()
+
+	wait.Until(func() {
+		targetFirewallState := fc.fwCache.firewall.state
+		fm := *fc.firewallManager
+
+		currentFirewallState, err := fm.Get(context.Background())
+		if err != nil {
+			klog.Errorf("failed to get worker firewall: %s", err)
+		}
+
+		if &currentFirewallState != nil {
+			if cmp.Equal(currentFirewallState, targetFirewallState) {
+				return
+			}
+		}
+		err = fc.onServiceChange(ctx)
+		if err != nil {
+			klog.Errorf("Failed to reconcile worker firewall: %s", err)
+		}
+
+	}, 5*time.Minute, stopCh)
+}
+
+// Get returns the current CCM worker firewall representation.
+func (fm *firewallManagerOp) Get(ctx context.Context) (*godo.Firewall, error) {
+	// check cache and query the API firewall service to get firewall ID, if it exists. Return it. If not, continue.
+	if notNil(fm) {
+		fw, resp, err := fm.client.Firewalls.Get(ctx, fm.fwCache.firewall.state.ID)
+		if err != nil {
+			klog.Errorf("failed to retrieve firewall by ID from local cache")
+		}
+		if resp != nil && resp.StatusCode == 404 {
+			klog.Infof("unable to retrieve firewall by ID because it no longer exists")
+		}
+		if fw != nil {
+			return fw, nil
+		}
+	}
+
+	// iterate through firewall API provided list and return the firewall with the matching firewall name.
+	firewallList, err := allFirewallList(ctx, fm.client)
```

Comment:
For accounts with a large number of firewalls (rare case but could happen), we'd load all of them into memory first before filtering. It'd be more efficient to change `allFirewallList` to accept a boolean filtering function that takes a `*godo.Firewall` and returns it if the filtering function returns true.

Reply:
Good suggestion. Let me know if what I update the code to makes sense.